roles/upgrade/pre-upgrade/tasks/drain.yml

```diff
@@ -0,0 +1,40 @@
+---
+- name: Group of tasks for a drain node with retries and notifications
+  block:
+    - name: Increment the retry count
+      set_fact:
+        notification_retry_count: "{{ 0 if notification_retry_count is undefined else notification_retry_count | int + 1 }}"
+
+    - name: Drain node
+      command: >-
+        {{ bin_dir }}/kubectl drain
+        --force
+        --ignore-daemonsets
+        --grace-period {{ drain_grace_period }}
+        --timeout {{ drain_timeout }}
+        --delete-local-data {{ kube_override_hostname|default(inventory_hostname) }}
+        {% if drain_pod_selector %}--pod-selector '{{ drain_pod_selector }}'{% endif %}
+      when:
+        - drain_nodes
+      delegate_to: "{{ groups['kube-master'][0] }}"
+  rescue:
+    - name: Check if retries left
+      fail:
+        msg: "Node can't drain pods. Please take a look on it {{ inventory_hostname }}"
+      when: notification_retry_count | int == drain_max_reties or not drain_retry_enabled
+
+    - name: Send Microsoft Teams notification
```

Comment:
I'm really not sure about adding this to the codebase... this open the way to slack, MM etc.. not sure we want to support every connector.
Maybe something could be added in the documentation or contribs and reference somehow ?

Reply:
Agree. this should be in a contrib folder or generic enough to support pluggable IM systems